### PR TITLE
Image2neibs infer shape

### PR DESCRIPTION
**Pull request by yaoli**
implemented infer_shape for image2neibs op

### CODE REVIEW

**Comment by nouiz** [comment]
You must clean this PR before merging it. I think it would be easier to cherry pick the right commit(s) on the current Theano master.

File: theano/tensor/nnet/neighbours.py [hunk 1]
```diff
@@ -80,11 +80,14 @@ def make_node(self, ten4, neib_shape, neib_step=None):
...
         else:
             neib_step = T.as_tensor_variable(neib_step)
-
+        
         assert ten4.ndim == 4
         assert neib_shape.ndim == 1
         assert neib_step.ndim == 1
 
+        self.neib_shape = neib_shape
+        self.neib_step = neib_step
```
Comment (nouiz) [inline theano/tensor/nnet/neighbours.py]: This is not good. You can't keep in the op a link to the Theano variables for the Apply. The same op is shared between many apply. There would be collision. Remove that change. In the infer_shape, you can access them throught the node parameter.

File: theano/tensor/nnet/neighbours.py [hunk 2]
```diff
@@ -208,7 +211,29 @@ def CEIL_INTDIV(a, b):
...
+                                
+    def infer_shape(self, node, input_shape):
+        in_shape = input_shape[0]
+        def CEIL_INTDIV(a, b):
+            if a % b:
+                return (a // b) + 1
+            else:
+                return a // b
+        c, d = self.neib_shape
+        step_x, step_y = self.neib_step
```
Comment (nouiz) [inline theano/tensor/nnet/neighbours.py]: use node.inputs[2]

File: theano/tensor/nnet/tests/test_neighbours.py [hunk 3]
```diff
@@ -366,6 +366,37 @@ def speed_neibs_wrap_centered(self):
...
+        images = numpy.ones(shape).astype('float32')
+        x = T.ftensor4()
+        f = self._compile_and_check([x],
+                                    [images2neibs(
+                                        x, neib_shape=(2,2),
+                                        mode='valid')],
+                                    [images],
+                                    Images2Neibs
+                                    )
+        shape = (100, 40, 5, 5)
```
Comment (nouiz) [inline theano/tensor/nnet/tests/test_neighbours.py]: Test ignore border with this shape (100, 40, 6, 6). This is another corner case.

File: theano/tensor/nnet/tests/test_neighbours.py [hunk 4]
```diff
@@ -366,6 +366,37 @@ def speed_neibs_wrap_centered(self):
...
+        images = numpy.ones(shape).astype('float32')
+        x = T.ftensor4()
+        f = self._compile_and_check([x],
+                                    [images2neibs(
+                                        x, neib_shape=(2,2),
+                                        mode='ignore_borders')],
+                                    [images],
+                                    Images2Neibs
+                                    )
+        shape = (100, 40, 5, 5)
```
Comment (nouiz) [inline theano/tensor/nnet/tests/test_neighbours.py]: idem, test with 100, 40, 6, 6

File: theano/tensor/nnet/tests/test_neighbours.py (hunk 3 above)
Comment (nouiz) [inline theano/tensor/nnet/tests/test_neighbours.py]: Also tests with neib_shape (2, 3).

File: theano/tensor/nnet/neighbours.py [hunk 5]
```diff
@@ -208,7 +211,29 @@ def CEIL_INTDIV(a, b):
...
-
+                                
+    def infer_shape(self, node, input_shape):
+        in_shape = input_shape[0]
+        def CEIL_INTDIV(a, b):
+            if a % b:
+                return (a // b) + 1
+            else:
+                return a // b
+        c, d = self.neib_shape
```
Comment (nouiz) [inline theano/tensor/nnet/neighbours.py]: use node.inputs[1]

**Comment by nouiz** [comment]
@yaoli do not forget to make a comment when you update a PR. That way we know about the change and can give feedback faster.

File: theano/tensor/nnet/neighbours.py [hunk 6]
```diff
@@ -80,7 +80,7 @@ def make_node(self, ten4, neib_shape, neib_step=None):
             neib_step = neib_shape
         else:
             neib_step = T.as_tensor_variable(neib_step)
-
+
```
Comment (abergeron) [inline theano/tensor/nnet/neighbours.py]: Don't add spaces in blank lines.

File: theano/tensor/nnet/neighbours.py [hunk 7]
```diff
@@ -208,7 +208,29 @@ def CEIL_INTDIV(a, b):
                                 z_col = j + d * i
 
                                 z[0][z_row, z_col] = ten4[n, s, ten4_2, ten4_3]
-
+                                
+    def infer_shape(self, node, input_shape):
+        in_shape = input_shape[0]
+        def CEIL_INTDIV(a, b):
```
Comment (abergeron) [inline theano/tensor/nnet/neighbours.py]: This is not suitable for symbolic variables.  Use theano.tensor.ceil_intdiv instead

**Comment by abergeron** [comment]
If you fix my two comments and address Fred's concerns with the tests, I will consider this finished.

**Comment by yaoli** [comment]
addressed final issues

**Comment by nouiz** [comment]
If I ran the tests, I have this warning many times with different shapes:

WARNING (theano.tests.unittest_tools): While testing the shape inference, we received an input with a shape that has some repeated values: (100, 40, 5, 5), like a square matrix. This makes it impossible to check if the values for these dimensions have been correctly used, or if they have been mixed up.

Can you fix it?

**Comment by yaoli** [comment]
All should be fixed.

**Comment by nouiz** [comment]
thanks
